File: pwa/src/lib/protocol/spake.ts
import { p256 } from "@noble/curves/p256.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { argon2id } from "hash-wasm";
import { concat, encStr, hexToBytes, lenPref } from "./bytes.ts";
import { hmacSha256, pairfobHKDF } from "./kdf.ts";

export const CONTEXT = "pairfob-v1/spake2plus";

const M = hexToBytes(
  "04886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f5ff355163e43ce224e0b0e65ff02ac8e5c7be09419c785e0ca547d55a12e2d20",
);
const N = hexToBytes(
  "04d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b4907d60aa6bfade45008a636337f5168c64d9bd36034808cd564490b1e656edbe7",
);

const ORDER = p256.CURVE.n;

function os2ipMod(b: Uint8Array): bigint {
  let n = 0n;
  for (const x of b) n = (n << 8n) | BigInt(x);
  return n % ORDER;
}

function pad32(n: bigint): Uint8Array {
  let hex = n.toString(16);
  if (hex.length % 2) hex = "0" + hex;
  const raw = hexToBytes(hex);
  const out = new Uint8Array(32);
  out.set(raw, 32 - raw.length);
  return out;
}

function pointMul(p: Uint8Array, k: bigint) {
  return p256.ProjectivePoint.fromHex(p).multiply(k);
}

function pointAdd(a: ReturnType<typeof pointMul>, b: ReturnType<typeof pointMul>) {
  return a.add(b);
}

function uncomp(p: ReturnType<typeof pointMul>): Uint8Array {
  return p.toRawBytes(false);
}

export interface Record {
  w0: bigint;
  w1: bigint;
  L: Uint8Array;
}

export async function deriveRecord(normalizedS: string, daemonId: string, pairRefHex: string): Promise<Record> {
  const saltFull = sha256(
    concat(encStr("pairfob-v1/pake-salt"), encStr(daemonId), encStr(pairRefHex)),
  );
  const salt = new Uint8Array(saltFull.subarray(0, 16));
  const expanded = await argon2id({
    password: normalizedS,
    salt,
    parallelism: 1,
    iterations: 3,
    memorySize: 65536,
    hashLength: 80,
    outputType: "binary",
  });
  const w0 = os2ipMod(expanded.slice(0, 40));
  const w1 = os2ipMod(expanded.slice(40, 80));
  const L = uncomp(p256.ProjectivePoint.BASE.multiply(w1));
  return { w0, w1, L };
}

export interface SpakeKeys {
  kShared: Uint8Array;
  confirmP: Uint8Array;
  confirmV: Uint8Array;
}

export function computeKeys(
  ctx: string,
  idP: string,
  idV: string,
  shareP: Uint8Array,
  shareV: Uint8Array,
  Z: Uint8Array,
  V: Uint8Array,
  w0: bigint,
): SpakeKeys {
  const tt = concat(
    encStr(ctx),
    encStr(idP),
    encStr(idV),
    lenPref(M),
    lenPref(N),
    lenPref(shareP),
    lenPref(shareV),
    lenPref(Z),
    lenPref(V),
    lenPref(pad32(w0)),
  );
  const kMain = sha256(tt);
  const conf = pairfobHKDF(kMain, null, new TextEncoder().encode("ConfirmationKeys"), 64);
  return {
    kShared: pairfobHKDF(kMain, null, new TextEncoder().encode("SharedKey"), 32),
    confirmP: hmacSha256(conf.slice(0, 32), shareV),
    confirmV: hmacSha256(conf.slice(32), shareP),
  };
}

export class Prover {
  x = 0n;
  shareP = new Uint8Array();
  constructor(
    public rec: Record,
    public idP: string,
    public idV: string,
    public ctx = CONTEXT,
  ) {}

  start(x?: bigint): Uint8Array {
    const bytes = new Uint8Array(32);
    crypto.getRandomValues(bytes);
    this.x = x ?? os2ipMod(bytes);
    if (this.x === 0n) this.x = 1n;
    const wM = pointMul(M, this.rec.w0);
    const xP = p256.ProjectivePoint.BASE.multiply(this.x);
    this.shareP = new Uint8Array(uncomp(pointAdd(xP, wM)));
    return this.shareP;
  }

  finish(shareV: Uint8Array): SpakeKeys {
    const Y = p256.ProjectivePoint.fromHex(shareV);
    const wN = pointMul(N, this.rec.w0);
    const tmp = Y.add(wN.negate());
    const Z = uncomp(tmp.multiply(this.x));
    const V = uncomp(tmp.multiply(this.rec.w1));
    return computeKeys(this.ctx, this.idP, this.idV, this.shareP, shareV, Z, V, this.rec.w0);
  }
}

export function idProver(pairRefHex: string): string {
  return "phone" + pairRefHex;
}

File: internal/crypto/spake2plus/spake.go
// Package spake2plus implements RFC 9383 SPAKE2+-P256-SHA256-HKDF-SHA256-HMAC-SHA256
// with Pairfob Argon2id → (w0,w1,L) (no second RFC PBKDF).
package spake2plus

import (
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"errors"
	"math/big"

	"golang.org/x/crypto/argon2"

	"pairfob/internal/crypto/canon"
	"pairfob/internal/crypto/hkdfk"
)

var (
	curve = elliptic.P256()
	order = curve.Params().N
	// Uncompressed M, N matching RFC 9383 Appendix C transcript encodings.
	mUncompressed = mustHex("04886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f5ff355163e43ce224e0b0e65ff02ac8e5c7be09419c785e0ca547d55a12e2d20")
	nUncompressed = mustHex("04d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b4907d60aa6bfade45008a636337f5168c64d9bd36034808cd564490b1e656edbe7")
)

const Context = "pairfob-v1/spake2plus"

type Record struct {
	W0, W1 *big.Int
	L      []byte // uncompressed 65
}

func DeriveRecord(normalizedS, daemonID, pairRefHex string) Record {
	saltFull := sha256.Sum256(canon.Concat(
		canon.EncStr("pairfob-v1/pake-salt"),
		canon.EncStr(daemonID),
		canon.EncStr(pairRefHex),
	))
	expanded := argon2.IDKey([]byte(normalizedS), saltFull[:16], 3, 64*1024, 1, 80)
	w0 := os2ipMod(expanded[0:40])
	w1 := os2ipMod(expanded[40:80])
	lx, ly := curve.ScalarBaseMult(pad32(w1))
	return Record{W0: w0, W1: w1, L: elliptic.Marshal(curve, lx, ly)}
}

type Keys struct {
	KShared, ConfirmP, ConfirmV, KMain []byte
}

func ComputeKeys(ctx, idP, idV string, shareP, shareV, Z, V []byte, w0 *big.Int) Keys {
	if ctx == "" {
		ctx = Context
	}
	tt := canon.Concat(
		canon.EncStr(ctx),
		canon.EncStr(idP),
		canon.EncStr(idV),
		lenPref(mUncompressed),
		lenPref(nUncompressed),
		lenPref(shareP),
		lenPref(shareV),
		lenPref(Z),
		lenPref(V),
		lenPref(pad32(w0)),
	)
	sum := sha256.Sum256(tt)
	kMain := sum[:]
	conf := hkdfk.HKDF(kMain, nil, []byte("ConfirmationKeys"), 64)
	return Keys{
		KMain:    kMain,
		KShared:  hkdfk.HKDF(kMain, nil, []byte("SharedKey"), 32),
		ConfirmP: hkdfk.HMACSHA256(conf[:32], shareV),
		ConfirmV: hkdfk.HMACSHA256(conf[32:], shareP),
	}
}

func ConfirmEqual(a, b []byte) bool {
	return subtle.ConstantTimeCompare(a, b) == 1
}

type Prover struct {
	W0, W1            *big.Int
	x                 *big.Int
	ShareP            []byte
	idP, idV, context string
}

func NewProver(rec Record, idProver, idVerifier, context string) *Prover {
	if context == "" {
		context = Context
	}
	return &Prover{W0: rec.W0, W1: rec.W1, idP: idProver, idV: idVerifier, context: context}
}

func (p *Prover) Start() []byte { return p.StartWithX(randomScalar()) }

func (p *Prover) StartWithX(x *big.Int) []byte {
	p.x = x
	mx, my := mustUnmarshal(mUncompressed)
	wMx, wMy := curve.ScalarMult(mx, my, pad32(p.W0))
	bx, by := curve.ScalarBaseMult(pad32(x))
	sx, sy := curve.Add(bx, by, wMx, wMy)
	p.ShareP = elliptic.Marshal(curve, sx, sy)
	return p.ShareP
}

func (p *Prover) Finish(shareV []byte) (Keys, error) {
	Yx, Yy, err := unmarshal(shareV)
	if err != nil {
		return Keys{}, err
	}
	nx, ny := mustUnmarshal(nUncompressed)
	wNx, wNy := curve.ScalarMult(nx, ny, pad32(p.W0))
	nNx, nNy := neg(wNx, wNy)
	tx, ty := curve.Add(Yx, Yy, nNx, nNy)
	Zx, Zy := curve.ScalarMult(tx, ty, pad32(p.x))
	Vx, Vy := curve.ScalarMult(tx, ty, pad32(p.W1))
	return ComputeKeys(p.context, p.idP, p.idV, p.ShareP, shareV,
		elliptic.Marshal(curve, Zx, Zy), elliptic.Marshal(curve, Vx, Vy), p.W0), nil
}

type Verifier struct {
	W0                *big.Int
	L                 []byte
	y                 *big.Int
	ShareV            []byte
	idP, idV, context string
}

func NewVerifier(rec Record, idProver, idVerifier, context string) *Verifier {
	if context == "" {
		context = Context
	}
	return &Verifier{W0: rec.W0, L: rec.L, idP: idProver, idV: idVerifier, context: context}
}

func (v *Verifier) Finish(shareP []byte) (shareV []byte, keys Keys, err error) {
	return v.FinishWithY(shareP, randomScalar())
}

func (v *Verifier) FinishWithY(shareP []byte, y *big.Int) (shareV []byte, keys Keys, err error) {
	Xx, Xy, err := unmarshal(shareP)
	if err != nil {
		return nil, Keys{}, err
	}
	v.y = y
	nx, ny := mustUnmarshal(nUncompressed)
	wNx, wNy := curve.ScalarMult(nx, ny, pad32(v.W0))
	bx, by := curve.ScalarBaseMult(pad32(y))
	Yx, Yy := curve.Add(bx, by, wNx, wNy)
	v.ShareV = elliptic.Marshal(curve, Yx, Yy)

	mx, my := mustUnmarshal(mUncompressed)
	wMx, wMy := curve.ScalarMult(mx, my, pad32(v.W0))
	nMx, nMy := neg(wMx, wMy)
	tx, ty := curve.Add(Xx, Xy, nMx, nMy)
	Zx, Zy := curve.ScalarMult(tx, ty, pad32(y))
	Lx, Ly, err := unmarshal(v.L)
	if err != nil {
		return nil, Keys{}, err
	}
	Vx, Vy := curve.ScalarMult(Lx, Ly, pad32(y))
	keys = ComputeKeys(v.context, v.idP, v.idV, shareP, v.ShareV,
		elliptic.Marshal(curve, Zx, Zy), elliptic.Marshal(curve, Vx, Vy), v.W0)
	return v.ShareV, keys, nil
}

func IdProver(pairRefHex string) string { return "phone" + pairRefHex }

func os2ipMod(b []byte) *big.Int {
	n := new(big.Int).SetBytes(b)
	return n.Mod(n, order)
}

func pad32(n *big.Int) []byte {
	b := n.Bytes()
	if len(b) > 32 {
		b = b[len(b)-32:]
	}
	out := make([]byte, 32)
	copy(out[32-len(b):], b)
	return out
}

func unmarshal(b []byte) (x, y *big.Int, err error) {
	x, y = elliptic.Unmarshal(curve, b)
	if x == nil {
		return nil, nil, errors.New("invalid P-256 point")
	}
	return x, y, nil
}

func mustUnmarshal(b []byte) (x, y *big.Int) {
	x, y, err := unmarshal(b)
	if err != nil {
		panic(err)
	}
	return x, y
}

func neg(x, y *big.Int) (*big.Int, *big.Int) {
	yy := new(big.Int).Neg(y)
	yy.Mod(yy, curve.Params().P)
	return new(big.Int).Set(x), yy
}

func randomScalar() *big.Int {
	for {
		k, err := rand.Int(rand.Reader, order)
		if err != nil {
			panic(err)
		}
		if k.Sign() != 0 {
			return k
		}
	}
}

func lenPref(b []byte) []byte {
	out := make([]byte, 8+len(b))
	n := uint64(len(b))
	for i := 0; i < 8; i++ {
		out[i] = byte(n >> (8 * i))
	}
	copy(out[8:], b)
	return out
}

func mustHex(s string) []byte {
	b := make([]byte, len(s)/2)
	for i := 0; i < len(b); i++ {
		var v byte
		for _, c := range []byte{s[2*i], s[2*i+1]} {
			v <<= 4
			switch {
			case c >= '0' && c <= '9':
				v |= c - '0'
			case c >= 'a' && c <= 'f':
				v |= c - 'a' + 10
			case c >= 'A' && c <= 'F':
				v |= c - 'A' + 10
			}
		}
		b[i] = v
	}
	return b
}

func ScalarFromHex(s string) *big.Int {
	n := new(big.Int)
	n.SetString(s, 16)
	return n
}

func RecordFromHex(w0, w1, L string) Record {
	return Record{W0: ScalarFromHex(w0), W1: ScalarFromHex(w1), L: mustHex(L)}
}
